apps/emqx/src/emqx_channel.erl

```diff
@@ -1137,16 +1137,14 @@ handle_call(
     kick,
     Channel = #channel{
         conn_state = ConnState,
-        will_msg = WillMsg,
-        clientinfo = ClientInfo,
         conninfo = #{proto_ver := ProtoVer}
     }
 ) ->
-    (WillMsg =/= undefined) andalso publish_will_msg(ClientInfo, WillMsg),
+    Channel0 = maybe_publish_will_msg(kick, Channel),
```

Comment:
Q: Also kinda confused why this case needs separate handling at the time of `handle_call/2`? The first impression is the same would be achieved by just removing this code and separate clause with `terminate({shutdown, kicked}, Channel)` from `terminate/2`.

Reply:
I am not 100% sure if kick also means delete session here. 
I will double check in the spec.